airflow/providers/google/cloud/triggers/dataflow.py

```diff
@@ -142,3 +152,508 @@ def _get_async_hook(self) -> AsyncDataflowHook:
+        If set as a string, the account must grant the originating account
+        the Service Account Token Creator IAM role.
+        If set as a sequence, the identities from the list must grant
+        Service Account Token Creator IAM role to the directly preceding identity, with first
+        account from the list granting this role to the originating account (templated).
+    :param fail_on_terminal_state: If set to True the trigger will yield a TriggerEvent with
+        error status if the job reaches a terminal state.
+    """
+
+    def __init__(
+        self,
+        job_id: str,
+        project_id: str | None,
+        location: str = DEFAULT_DATAFLOW_LOCATION,
+        gcp_conn_id: str = "google_cloud_default",
+        poll_sleep: int = 10,
+        impersonation_chain: str | Sequence[str] | None = None,
+        fail_on_terminal_state: bool = True,
+    ):
+        super().__init__()
+        self.project_id = project_id
+        self.job_id = job_id
+        self.location = location
+        self.gcp_conn_id = gcp_conn_id
+        self.poll_sleep = poll_sleep
+        self.impersonation_chain = impersonation_chain
+        self.fail_on_terminal_state = fail_on_terminal_state
+
+    def serialize(self) -> tuple[str, dict[str, Any]]:
+        """Serialize class arguments and classpath."""
+        return (
+            "airflow.providers.google.cloud.triggers.dataflow.DataflowJobMessagesTrigger",
+            {
+                "project_id": self.project_id,
+                "job_id": self.job_id,
+                "location": self.location,
+                "gcp_conn_id": self.gcp_conn_id,
+                "poll_sleep": self.poll_sleep,
+                "impersonation_chain": self.impersonation_chain,
+                "fail_on_terminal_state": self.fail_on_terminal_state,
+            },
+        )
+
+    async def run(self):
+        """
+        Loop until a terminal job status or any job messages are returned.
+
+        Yields a TriggerEvent with success status, if the client returns any job messages
+        and fail_on_terminal_state attribute is False.
+
+        Yields a TriggerEvent with error status, if the client returns a job status with
+        a terminal state value and fail_on_terminal_state attribute is True.
+
+        Yields a TriggerEvent with error status, if any exception is raised while looping.
+
+        In any other case the Trigger will wait for a specified amount of time
+        stored in self.poll_sleep variable.
+        """
+        try:
+            while True:
+                job_status = await self.async_hook.get_job_status(
+                    job_id=self.job_id,
+                    project_id=self.project_id,
+                    location=self.location,
+                )
+                job_messages = await self.list_job_messages()
+                if self.loop_must_fail_when_job_is_in_terminal_state(job_status):
```

Comment:
Is there any reason we need to name the method name this long? Will something like `job_is_in_termianl_state` work?

Reply:
I just found out they might be different, but wondering do we need to make them a method? It seems they're not reused.

Reply:
@Lee-W in one of your previous comments you suggested to move the compound if statements (`if self.fail_on_terminal_state and job_status.name in DataflowJobStatus.TERMINAL_STATES`) to a function. Given that one of the statements was using `self`, I decided to move it into a method, but it's not like I spent too much time thinking about it:) 
To be fair, most of the trigger's methods are not reused, they're just called multiple times in a loop, and the same is true for this one. Do you think it would be better to move this code outside of the class?

As for the name, I chose a longer one because I had a hard time expressing what this method really does. Something like `job_is_in_terminal_state` didn't convey that the sensor must exit the loop only if `self.fail_on_terminal_state` is set to *True*. If it is set to *False*, then reaching the terminal state should not affect anything.

Reply:
> Do you think it would be better to move this code outside of the class?

If they're the same thing across different classes, I think we could move it outside of class or make another base class with this method.

> As for the name, I chose a longer one because I had a hard time expressing what this method really does. 

Hmmm... if that's the case, I think we probably should stick with the original if statement. I feel like the function name does not enhance the readbility 🤔

Reply:
Returned to the if statement.